### PR TITLE
[working] Rados bucket support

### PR DESCRIPTION
**Pull request by comzyh**

### CODE REVIEW

File: bucket/rados.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	name   string
+	offset uint64
+}
+
+// Close close reader
+func (reader RadosObjectReader) Close() error {
+	return nil
+}
+
+// Read close reader
```
Comment (xiang90) [inline bucket/rados.go]: wrong comment

File: bucket/rados.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+func (reader RadosObjectWriter) Close() error {
+	return nil
+}
+
+func (writer RadosObjectWriter) Write(data []byte) (int, error) {
+	err := writer.bucket.ioctx.Write(writer.bucket.prefix+writer.name, data, writer.offset)
+	if err != nil {
+		writer.offset += uint64(len(data))
+		return 0, err
+	} else {
```
Comment (xiang90) [inline bucket/rados.go]: remove this eles

File: bucket/rados.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+func NewRadosBucket(mons, secret, pool, prefix string) (bk Bucket, err error) {
+	conn, err := rados.NewConn()
+	conn.SetConfigOption("mon_host", mons)
+	conn.SetConfigOption("key", secret)
+	ioctx, err := conn.OpenIOContext(pool)
+	if err != nil {
+		log.Printf("Cannot open pool %s: %v", pool, err)
```
Comment (xiang90) [inline bucket/rados.go]: this should be logged at caller side.

File: bucket/rados.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		writer.offset += uint64(len(data))
+		return 0, err
+	} else {
+		return len(data), nil
+	}
+}
+
+func NewRadosBucket(mons, secret, pool, prefix string) (bk Bucket, err error) {
+	conn, err := rados.NewConn()
```
Comment (xiang90) [inline bucket/rados.go]: check error here.

File: bucket/bucket.go [hunk 5]
```diff
@@ -24,6 +24,38 @@ type Bucket interface {
...
+		if !ok {
+			return nil, fmt.Errorf("pool is not provided")
+		}
+		prefix, ok := config["prefix"]
+		if !ok {
+			prefix = ""
+		}
+		return NewRadosBucket(mons.(string), secret.(string), pool.(string), prefix.(string))
+	}
+	return nil, fmt.Errorf("Unknown bucketType %s", bucketType)
```
Comment (xiang90) [inline bucket/bucket.go]: unknown bucket type

Comment (xiang90) [inline bucket/bucket.go]: probably put this into default in switch instead.

File: bucket/rados.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+func (reader *RadosObjectReader) Close() error {
+	return nil
+}
+
+// Read read from object
+func (reader *RadosObjectReader) Read(p []byte) (n int, err error) {
+	n, err = reader.bucket.ioctx.Read(reader.bucket.prefix+reader.name, p, reader.offset)
+	reader.offset += uint64(n)
+	if n == 0 {
+		return 0, io.EOF
```
Comment (xiang90) [inline bucket/rados.go]: this is wrong. we might overwrite the actual error when n is 0. for example when we get 0, "connection refused"

File: bucket/rados.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	conn, err := rados.NewConn()
+	if err != nil {
+		return nil, err
+	}
+	conn.SetConfigOption("mon_host", mons)
+	conn.SetConfigOption("key", secret)
+	err = conn.Connect()
+	if err != nil {
+		log.Println("Failed to connect ceph:", err)
+		return
```
Comment (xiang90) [inline bucket/rados.go]: kill naked return

return nil,err

**Comment by xiang90** [comment]
lgtm

File: bucket/rados.go (hunk 6 above)
Comment (comzyh) [inline bucket/rados.go]: fixed

File: cmd/ceph/main.go [hunk 8]
```diff
@@ -1,65 +1,67 @@
...
+		log.Fatalf("Can't open %s for write: %v", objName, err)
 	}
-
-	err = conn.Connect()
-	if err != nil {
-		fmt.Println("Failed to connect:", err)
-		return
+	lines := []string{
+		"Hello World\n",
+		"This is the firsst Line,\n",
```
Comment (Lcch) [inline cmd/ceph/main.go]: typo: this is the first line

File: cmd/master/main.go [hunk 9]
```diff
@@ -10,6 +10,7 @@ import (
 	"os"
 	"strings"
 
+	"github.com/naturali/kmr/job"
 	"github.com/naturali/kmr/master"
```
Comment (Lcch) [inline cmd/master/main.go]: remove the blank line

**Pull request merged**